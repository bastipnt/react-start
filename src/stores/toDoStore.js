// @flow
import { observable } from 'mobx';

class ToDo implements TODO {
  @observable text;
  @observable inEdit;
  @observable completed;
  createdAt;
  changedAt;

  constructor(text: string) {
    this.text = text;
    this.createdAt = Date.now();
    this.inEdit = false;
    this.completed = false;
  }

  change = (text: string) => {
    this.text = text;
    this.changedAt = Date.now();
    this.inEdit = false;
    this.completed = false;
  }

  edit = () => { this.inEdit = true; }

  cancelEdit = () => { this.inEdit = false; }

  toggleCompleted = () => { this.completed = !this.completed; }
}

class ToDoStore implements TODOSTORE {
  @observable toDos = []

  createToDo = (text: string) => { if (text) this.toDos.push(new ToDo(text)); }

  deleteToDo = (toDo: TODO) => { this.toDos.remove(toDo); }
}

export default new ToDoStore();
