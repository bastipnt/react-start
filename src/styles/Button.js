import styled from 'styled-components';

const Button = styled.button`
  color: blue;
  border: 2px solid red;
`;

export default Button;
