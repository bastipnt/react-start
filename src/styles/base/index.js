export * from './Button';
export * from './Input';
export * from './List';
export * from './P';
export * from './Layout';
export * from './MD';
