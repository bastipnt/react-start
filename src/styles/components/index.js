export * from './ToDoList.styles';
export * from './Header.styles';
