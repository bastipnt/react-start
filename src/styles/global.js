import { injectGlobal } from 'styled-components';

/* eslint no-unused-expressions: 0 */
injectGlobal`
  html {
    box-sizing: border-box;
    -webkit-font-smoothing: antialiased;
    -webkit-tap-highlight-color: transparent;
    text-rendering: optimizeLegibility;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  html,
  body {
    height: 100%;
    width: 100%;
  }

  body {
    font-family: 'Helvetica Neue', Helvetica, Arial, sans-serif;
    color: #fff;
  }

  body.fontLoaded {
    font-family: 'Open Sans', 'Helvetica Neue', Helvetica, Arial, sans-serif;
  }

  #app {
    min-height: 100%;
    min-width: 100%;
    display: flex;
    flex-flow: row;
  }
`;
