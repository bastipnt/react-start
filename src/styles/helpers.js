// @flow
import CF from 'colors.js';

const validateHEX = (hex: string): string => {
  let hexWithoutHash = hex.match(/#(([a-f0-9]{3}){1,2})\b/i);
  if (!hexWithoutHash) hexWithoutHash = '000';
  else [, hexWithoutHash] = hexWithoutHash;
  const fullHex = hexWithoutHash.length < 6 ?
    hexWithoutHash[0] + hexWithoutHash[0] + hexWithoutHash[1] + hexWithoutHash[1] +
    hexWithoutHash[2] + hexWithoutHash[2] : hexWithoutHash;
  return `#${fullHex}`;
};

export const opacify = (hex: string = '#fff', opacity?: number = 1): string => {
  const validHex = validateHEX(hex);
  return `rgba(${CF.hex2rgb(validHex).a.toString()}, ${String(opacity)})`;
};
