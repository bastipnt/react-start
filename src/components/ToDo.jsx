// @flow
import React, { Component } from 'react';
import { observer } from 'mobx-react';
import { TLItem, TLEdit, TLShow, Checkbox, Button } from 'styles';

type Props = {
  toDo: TODO,
  deleteToDo: (toDo: TODO) => void,
}

@observer
class ToDo extends Component<Props> {
  handleDoubleClick = (): void => this.props.toDo.edit()

  handleKeyPress = (e: SyntheticEvent<HTMLInputElement>): void => {
    const { toDo } = this.props;
    const target = e.currentTarget;
    if (e.key === 'Enter') {
      toDo.change(target.value);
      target.value = '';
    }
  }

  handleBlur = (): void => {
    const { toDo } = this.props;
    toDo.cancelEdit();
  }

  handleCheckboxChange = (): void => {
    const { toDo } = this.props;
    toDo.toggleCompleted();
  }

  handleDeleteClick = (): void => {
    const { deleteToDo, toDo } = this.props;
    deleteToDo(toDo);
  }

  render = () => {
    const { text, inEdit, completed } = this.props.toDo;

    const editMode = (
      <TLEdit
        defaultValue={text}
        onKeyPress={this.handleKeyPress}
        onBlur={this.handleBlur}
        autoFocus
      />
    );

    const showMode = <TLShow onDoubleClick={this.handleDoubleClick}>{text}</TLShow>;

    return (
      <TLItem completed={completed}>
        <Checkbox
          key="checkbox"
          onChange={this.handleCheckboxChange}
          checked={completed}
        />
        {inEdit ? editMode : showMode}
        <Button key="delete" onClick={this.handleDeleteClick}>
          <i className="fa fa-times" />
        </Button>
      </TLItem>
    );
  }
}

export default ToDo;
