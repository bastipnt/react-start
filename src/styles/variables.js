// @flow
import { opacify } from './helpers';

export const color = {
  text: opacify('#70708D', 1),
  textLight: opacify('#70708D', 0.4),
  textSpecial: '#6064FA',
  textHeader: opacify('#fff', 0.8),
  background: opacify('#fff', 1),
  backgroundDark: opacify('#70708D', 0.1),
  backgroundSpecial: '#FDFDFF',
  backgroundButton: opacify('#fff', 0.5),
  divider: opacify('#ECF2F6', 1),
  dividerLight: opacify('#ECF2F6', 1),
  placeholder: opacify('#6064FA', 0.5),
  checkboxBorder: opacify('#ECF2F6', 1),
  checkboxBackground: opacify('#6064FA', 1),
  checkboxText: opacify('#fff', 1),
  button: opacify('#70708D', 1),
  header: opacify('#fff', 0.1),
};
