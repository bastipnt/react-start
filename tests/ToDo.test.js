import React from 'react';
import { shallow } from 'enzyme';

import { TLEdit, TLShow, Checkbox, Button } from 'styles';
import ToDo from '../src/components/ToDo';
import toDoStore from '../src/stores/toDoStore';

const text = 'Test';
let component;

describe('<ToDo />', () => {
  describe('content', () => {
    beforeEach(() => {
      toDoStore.createToDo(text);
      const props = { toDo: toDoStore.toDos[0], deleteToDo: toDoStore.deleteToDo };
      component = shallow(<ToDo {...props} />);
    });

    afterEach(() => {
      toDoStore.toDos = [];
    });

    it('renders correctly', () => {
      expect(component.getElements()).toMatchSnapshot();
      expect(component.find(TLShow).dive().text()).toEqual(text);
    });

    it('shows an input field on doubleclick', () => {
      component.find(TLShow).simulate('doubleclick');
      expect(component.find(TLEdit).length).toBe(1);
    });

    it('removes the input field on blur', () => {
      component.find(TLShow).simulate('doubleclick');
      expect(component.find(TLEdit).length).toBe(1);
      component.find(TLEdit).simulate('blur');
      expect(component.find(TLEdit).length).toBe(0);
    });

    it('sets the toDo to done when clicking the checkbox', () => {
      expect(toDoStore.toDos[0].completed).toEqual(false);
      component.find(Checkbox).simulate('change', { target: { checked: true } });
      expect(toDoStore.toDos[0].completed).toEqual(true);
    });

    it('changes the text in edit mode when clicking "Enter"', () => {
      const textNew = 'TestNew';
      component.find(TLShow).simulate('doubleclick');
      component.find(TLEdit).simulate('keyPress', { key: 'Enter', currentTarget: { value: textNew } });
      expect(toDoStore.toDos[0].text).toEqual(textNew);
    });

    it('doesn`t change the text in edit mode when clicking any other key', () => {
      const textNew = 'TestNew';
      component.find(TLShow).simulate('doubleclick');
      component.find(TLEdit).simulate('keyPress', { key: 'A', currentTarget: { value: textNew } });
      expect(toDoStore.toDos[0].text).toEqual(text);
    });

    it('deletes the toDo when clicking the delete button', () => {
      component.find(Button).simulate('click');
      expect(toDoStore.toDos.length).toBe(0);
    });
  });
});
