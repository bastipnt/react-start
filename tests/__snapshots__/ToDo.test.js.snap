// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDo /> content renders correctly 1`] = `
Array [
  <List__LI
    completed={false}
  >
    <Input__Checkbox
      checked={false}
      onChange={[Function]}
    />
    <ToDoListstyles__TLShow
      onDoubleClick={[Function]}
    >
      Test
    </ToDoListstyles__TLShow>
    <Button
      onClick={[Function]}
    >
      <i
        className="fa fa-times"
      />
    </Button>
  </List__LI>,
]
`;
